---
title: "Microbial Innovations"
author: "Mario E. Muscarella, James P. O'Dwyer"
date: "Last updated on `r format(Sys.time(), '%d %B, %Y')`"
header-includes:
  - \usepackage{array}
  - \usepackage{graphics}
output: 
  pdf_document:
  fig_caption: true
---

## Initial Setup
```{r}
# Clear Environment and Set Working Directory
rm(list=ls())
setwd("~/GitHub/microbial-innovations/analyses")

# Add Basic Functions
se <- function(x, ...){sd(x, na.rm = TRUE)/sqrt(length(na.omit(x)))}
ci <- function(x, ...){1.96 * sd(x,na.rm = TRUE)}

# Load Packages
require("ape"); require("png"); require("grid")
require("ggtree")
require("phytools")
library("colorspace")
require("simecol")
require("picante")
require("tidyr")

# Load Source Functions
source("../bin/MothurTools.R")
source("../bin/di2multi2.R")
source("../bin/TraitMap.R")

# Define Simple Functions
ggplotColours <- function(n = 6, h = c(0, 360) + 15){
  if ((diff(h) %% 360) < 1) h[2] <- h[2] - 360/n
  hcl(h = (seq(h[1], h[2], length = n)), c = 100, l = 65)
}

ttest <- function(reg, coefnum, val){
  co <- coef(summary(reg))
  tstat <- (co[coefnum,1]-val)/co[coefnum,2]
  pstat <- 2 * pt(abs(tstat), reg$df.residual, lower.tail = FALSE)
  return(list = c(t = tstat, df = reg$df.residual, p =  pstat))
}

# Save Default Plot Settings
opar <- par(no.readonly = TRUE)  # Saves plot defaults

# Run All?
run.all <- FALSE
```

## Import Tree Files
```{r}
GG <- read.tree("../data/LTPs123.tree")
NTL <- read.tree("../data/NTL.tree")
RDP <- read.tree("../data/RDP.tree")
```

## Import Taxonomy Information
```{r}
LTP.tax <- read.tax("../data/LTPs123_unique.nr_v123.wang.taxonomy", format = "rdp", col.tax = 2, tax.levels = 6)
NTL.tax <- read.tax("../data/nmicrobiol201648_s7.pds.wang.taxonomy", format = "rdp", col.tax = 2, tax.levels = 6)
RDP.tax <- read.tax("../data/rdp_download_9752seqs.pds.wang.taxonomy", format = "rdp", col.tax = 2, tax.levels = 6)
```


# Edge Length Distribution
```{r, results = "hide"}
LTP.edges <- LTP$edge.length
NTL.edges <- NTL$edge.length
RDP.edges <- RDP$edge.length

png(filename="../figures/BranchLengths.png",
    width = 800, height = 1600, res = 96*2)

layout(matrix(1:3, 3, 1))
par(mar = c(5, 6, 3, 1) + 0.1, oma = c(1,1,1,1))

hist(log10(LTP.edges), main = "Silva: Living Tree Project",
     axes = F, xlab = "", ylab = "", xlim = c(-4, 1), col = "gray")

axis(1, at = c(-4, -3, -2, -1, 0, 1), 
     labels = expression(10^-4, 10^-3, 10^-2, 10^-1, 1, 10^1),  
     las = 1, lwd = 1.5)
axis(2, labels = T, las = 1, lwd = 1.5)

mtext("Edge Length", side = 1, cex = 1, line = 2.5)
mtext("Frequency", side = 2, cex = 1, line = 4)

hist(log10(NTL.edges), main = "Hug et al. 2016: New Tree of Life",
     axes = F, xlab = "", ylab = "", xlim = c(-4, 1), col = "gray")

axis(1, at = c(-4, -3, -2, -1, 0, 1), 
     labels = expression(10^-4, 10^-3, 10^-2, 10^-1, 1, 10^1),  
     las = 1, lwd = 1.5)
axis(2, labels = T, las = 1, lwd = 1.5)

mtext("Edge Length", side = 1, cex = 1, line = 2.5)
mtext("Frequency", side = 2, cex = 1, line = 4)

hist(log10(RDP.edges), main = "Ribosomal Datebase Project",
     axes = F, xlab = "", ylab = "", xlim = c(-4, 1), col = "gray")

axis(1, at = c(-4, -3, -2, -1, 0, 1), 
     labels = expression(10^-4, 10^-3, 10^-2, 10^-1, 1, 10^1),  
     las = 1, lwd = 1.5)
axis(2, labels = T, las = 1, lwd = 1.5)

mtext("Edge Length", side = 1, cex = 1, line = 2.5)
mtext("Frequency", side = 2, cex = 1, line = 4)


# Close Plot Device
dev.off()
graphics.off()
```
## Show Plot
```{r}
img <- readPNG("../figures/BranchLengths.png")
grid.raster(img)
```

## Edge Length vs Support
```{r}
node.tab <- data.frame(matrix(NA, LTP$Nnode, 5))
colnames(node.tab) <- c("node", "support", "edge_up", "edge_down1", "edge_down2")
node.tab$node <- as.numeric(c(1:LTP$Nnode))
node.tab$support <- LTP$node.label
node.tab[1,3] <- NA
for (i in 2:LTP$Nnode){
  edge.up <- which(LTP$edge[,2] == length(LTP$tip.label) + i)
  node.tab[i,3] <- LTP$edge.length[edge.up]
  edge.down <- which(LTP$edge[,1] == length(LTP$tip.label) + i)
  node.tab[i,4] <- LTP$edge.length[edge.down[1]]
  node.tab[i,5] <- LTP$edge.length[edge.down[2]]
}

node.tab$edge_down_min <- apply(cbind(node.tab$edge_down1, node.tab$edge_down2), 1, min)

scatterhist <- function(x, y, xlab="", ylab=""){
  zones=matrix(c(2,0,1,3), ncol=2, byrow=TRUE)
  par(oma = c(1,1,1,1))
  layout(zones, widths=c(4/5,1/5), heights=c(1/5,4/5))
  xhist = hist(x, plot=FALSE)
  yhist = hist(y, plot=FALSE)
  top = max(c(xhist$counts, yhist$counts))
  par(mar=c(5,5,1,1))
  plot(x,y, las = 1, xlab = "", ylab = "")
  mtext(xlab, side=1, line=3, outer=F, cex = 1.25)
  mtext(ylab, side=2, line=3, outer=F, cex = 1.25)
  par(mar=c(0,5,1,1))
  barplot(xhist$counts, axes=FALSE, space=0)
  par(mar=c(5,0,1,1))
  barplot(yhist$counts, axes=FALSE, space=0, horiz=TRUE)
}
```

```{r, results = "hide"}
png(filename="../figures/NodeSupport_LTP.png",
    width = 800, height = 800, res = 96*2)

# Upstream
scatterhist(x = log10(node.tab$edge_up), y = as.numeric(node.tab$support),
            xlab = "Upstream Edge Length", ylab = "Node Support")

# Downstream
scatterhist(log10(node.tab$edge_down_min), as.numeric(node.tab$support),
            xlab = "Downstream Edge Length", ylab = "Node Support")

mean(as.numeric(node.tab$support), na.rm = T)
mean(as.numeric(node.tab$support[log10(node.tab$edge_up) < -1.75]), na.rm = T)
mean(as.numeric(node.tab$support[log10(node.tab$edge_up) > 0.75]), na.rm = T)
mean(as.numeric(node.tab$support[log10(node.tab$edge_up) < -2.25]), na.rm = T)

# Close Plot Device
dev.off()
graphics.off()
```
## Show Plot
```{r}
img <- readPNG("../figures/NodeSupport_LTP.png")
grid.raster(img)
```

## Edge Length vs Support
```{r}
node.tab <- data.frame(matrix(NA, NTL$Nnode, 5))
colnames(node.tab) <- c("node", "support", "edge_up", "edge_down1", "edge_down2")
node.tab$node <- as.numeric(c(1:NTL$Nnode))
node.tab$support <- NTL$node.label
node.tab[1,3] <- NA
for (i in 2:NTL$Nnode){
  edge.up <- which(NTL$edge[,2] == length(NTL$tip.label) + i)
  node.tab[i,3] <- NTL$edge.length[edge.up]
  edge.down <- which(NTL$edge[,1] == length(NTL$tip.label) + i)
  node.tab[i,4] <- NTL$edge.length[edge.down[1]]
  node.tab[i,5] <- NTL$edge.length[edge.down[2]]
}

node.tab$edge_down_min <- apply(cbind(node.tab$edge_down1, node.tab$edge_down2), 1, min)
```

```{r, results = "hide"}
png(filename="../figures/NodeSupport_NTL.png",
    width = 800, height = 800, res = 96*2)

# Upstream
scatterhist(x = log10(node.tab$edge_up), y = as.numeric(node.tab$support),
            xlab = "Upstream Edge Length", ylab = "Node Support")

# Downstream
scatterhist(log10(node.tab$edge_down_min), as.numeric(node.tab$support),
            xlab = "Downstream Edge Length", ylab = "Node Support")

mean(as.numeric(node.tab$support), na.rm = T)
mean(as.numeric(node.tab$support[log10(node.tab$edge_up) < -1.75]), na.rm = T)
mean(as.numeric(node.tab$support[log10(node.tab$edge_up) > 0.75]), na.rm = T)
mean(as.numeric(node.tab$support[log10(node.tab$edge_up) < -2.25]), na.rm = T)

# Close Plot Device
dev.off()
graphics.off()
```
## Show Plot
```{r}
img <- readPNG("../figures/NodeSupport_NTL.png")
grid.raster(img)
```

## LTP Exploratory Tree with Domain Grouping
```{r, results = "hide"}
png(filename="../figures/LTP_explore.png",
    width = 1600, height = 1600, res = 96*2)

LTP.2 <- LTP
LTP.tax2 <- LTP.tax[match(LTP.2$tip.label, LTP.tax$OTU), ]
groupInfo <- split(LTP.2$tip.label, LTP.tax2$Domain)
livingTree <- groupOTU(LTP.2, groupInfo, group_name = "Domain")
levels(attributes(livingTree)$Domain) <- names(groupInfo)
ggtree(livingTree, aes(color = Domain), layout="rectangular") + 
  geom_text(aes(subset=!isTip, label = node), show.legend = FALSE,
            hjust=-0.1, vjust = -0.5, size = 2) + 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
  scale_color_manual(values=c(ggplotColours(n = 4)[2:3])) 

# Close Plot Device
dev.off()
graphics.off()

# Clean Up
```
## Show Plot
```{r}
img <- readPNG("../figures/LTP_explore.png")
grid.raster(img)
```

## Identify the LTP Node that seperates bacteria and archaea
```{r, results = "hide"}
png(filename="../figures/LTP_archaea.png",
    width = 1600, height = 1600, res = 96*2)
par(mar=c(1,1,1,3))
cp <- ggtree(livingTree, aes(color = Domain), layout="rectangular") %>% ggtree::collapse(node=18228)
cp + geom_point2(aes(subset=(node == 18227)), size=3, shape=18, 
                 colour=ggplotColours(n = 4)[1], fill = "gray80") + 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) +
  geom_text2(aes(subset=(node == 18228), label = "Archaea"), 
             show.legend = FALSE, hjust=-0.1, size = 3) + 
  guides(colour = guide_legend(override.aes = list(shape = NA, label = NA))) + 
  scale_color_manual(values=c(ggplotColours(n = 4)[2:3])) 

# Close Plot Device
dev.off()
graphics.off()

# Clean Up
rm(cp)
```
## Show Plot
```{r}
img <- readPNG("../figures/LTP_archaea.png")
grid.raster(img)
```

## Reroot LTP Tree
```{r, results = "hide"}
png(filename="../figures/LTP_root.png",
    width = 1600, height = 1600, res = 96*2)

livingTree.r <- midpoint.root(livingTree)
groupInfo <- split(LTP.2$tip.label, LTP.tax2$Domain)
livingTree.r2 <- groupOTU(livingTree.r, groupInfo, group_name = "Domain")
levels(attributes(livingTree.r2)$Domain) <- c("root", names(groupInfo))
rt <- ggtree(livingTree.r2, aes(color = Domain), layout="rectangular")
rt + geom_point2(aes(subset=(node == 11934)), size=4, shape=18, 
                 colour=ggplotColours(n = 4)[1], fill = "gray80") + 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
  scale_color_manual(values=c(ggplotColours(n = 4)[1:3])) 

# Close Plot Device
dev.off()
graphics.off()

# Clean Up
rm(rt)
```
## Show Plot
```{r}
img <- readPNG("../figures/LTP_root.png")
grid.raster(img)
```

## Export Rooted Tree
```{r}
write.tree(livingTree.r, file = "../data/LTP.rooted.tre")
write.nexus(livingTree.r, file = "../data/LTP.rooted.nex")
```

## LTP Exploratory Tree with Cyano Grouping
```{r, results = "hide"}
png(filename="../figures/LTP_Cyano.png",
    width = 1600, height = 1600, res = 96*2)

LTP.tax.full <- read.tax("../data/LTPs123_unique.full.taxonomy", col.tax = 2, tax.levels = 7)
LTP.tax.rooted <- LTP.tax.full[match(livingTree.r$tip.label, LTP.tax.full$OTU), ]
all.equal(as.factor(LTP.tax.rooted$OTU), as.factor(livingTree.r$tip.label))

LTP.cyano <- LTP.tax.rooted$Phylum

for (i in 1:length(LTP.cyano)){
  if (LTP.cyano[i] == "Cyanobacteria"){
    LTP.cyano[i] = "Cyanobacteria"
  } else {
    if (LTP.tax.rooted$Domain[i] == "Archaea"){
      LTP.cyano[i] = "Archaea"
    } else {
      LTP.cyano[i] = "Other Bacteria"
    }
  }
}

groupInfo <- split(livingTree.r$tip.label, LTP.cyano)

# Only 16 Cyanos?
length(grep("[C|c]yano*", LTP.tax.rooted$Phylum))

LTP.cy <- groupOTU(livingTree.r, groupInfo, group_name = "Cyano")
levels(attributes(LTP.cy)$Cyano) <- c("Root", names(groupInfo))


LT <- ggtree(LTP.cy, aes(color = Cyano, layout="rectangular"))

LT + geom_text(aes(subset=!isTip, label = node), show.legend = FALSE,
            hjust=-0.1, vjust = -0.5, size = 2) + 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
  scale_color_manual(values=c(ggplotColours(n = 4)[c(1,2,4,3)])) 

# Close Plot Device
dev.off()
graphics.off()

# Clean Up
#```
## Show Plot
#```{r}
img <- readPNG("../figures/LTP_Cyano.png")
grid.raster(img)
```

# Isolate Cyanobacteria and Identify the LTP Node that seperates cyanos
```{r, results = "hide"}
png(filename="../figures/LTP_Cyano2.png",
    width = 1600, height = 1600, res = 96*2)

viewClade(LT + #geom_tiplab() +
            geom_text(aes(subset=!isTip, label = node), show.legend = FALSE,
            hjust=-0.1, vjust = -0.5, size = 2) + 
            theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
            scale_color_manual(values=c(ggplotColours(n = 4)[c(3,4,2,1)])), 
          node=14077)

LTP.tax.full[which(LTP.tax.full$OTU %in% livingTree.Cyano$tip.label), ]

livingTree.Cyano <- extract.clade(livingTree.r, 14157)


# Close Plot Device
dev.off()
graphics.off()

# Clean Up
#```
## Show Plot
#```{r}
img <- readPNG("../figures/LTP_Cyano2.png")
grid.raster(img)
```

## Make Tree Ultrametric
I used treePL to make the tree ultrametric. My config file was pretty simple. 
It took way too long to get treePL to work. Maybe an alternative would be better.
It looks like there are some program in R. 
The most common are chronos, chronopl, chronoMPL. But I don't know how they compare.

The output file was `LTP.dated.tree`
```{r, results = "hide"}
LTP.um <- read.tree("../data/LTP.dated.tree")

png(filename="../figures/LTP_dated.png",
    width = 1600, height = 1600, res = 96*2)

livingTree.um <- groupOTU(LTP.um, groupInfo, group_name = "Domain")
levels(attributes(livingTree.um)$Domain) <- c("root", names(groupInfo))
um <- ggtree(livingTree.um, aes(color = Domain), layout="rectangular")
um + geom_point2(aes(subset=(node == 11934)), size=4, shape=18, 
                 colour=ggplotColours(n = 4)[1], fill = "gray80") + 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
  scale_color_manual(values=c(ggplotColours(n = 4)[1:3])) 

# Close Plot Device
dev.off()
graphics.off()

# Clear Up
rm(um)
```
## Show Plot
```{r}
img <- readPNG("../figures/LTP_dated.png")
grid.raster(img)
```

## Branch Lenght Comparision
```{r, results = "hide"}
png(filename="../figures/Branch_Comparision.png",
    width = 800, height = 800, res = 96*2)

reg <- livingTree.r2$edge.length
um <- livingTree.um$edge.length

# Are nodes the same?
sum(livingTree.r2$node.label != livingTree.um$node.label)

zero.branch <- which(reg == 0 | um == 0)
reg.l <- log10(reg[-zero.branch]) + 2.5
um.l <- log10(um[-zero.branch]) 

mod1 <- lm(um.l ~ reg.l)
summary(mod1)
ttest(mod1, 2, 1)

par(mar = c(6,6,1,1) + 0.1, oma = c(0.5,0.5,0.5,0.5))
plot(reg.l, um.l, 
     xlim = c(0, 3.5), ylim = c(-4, 4),
     axes=F, xlab = "", ylab = "")

# abline(mod1, from = -2.5, to = 1.5, lty = 2, lwd = 2, col = "red")
new <- data.frame(reg.l = seq(0, 3.2, 0.1))
lines(x = t(new), y = predict(mod1, new), lty = 2, lwd = 2, col = "red")
clip(min(reg.l), max(reg.l), min(um.l), max(um.l))
abline(mod1$coefficients[1], 1, lty = 3, lwd = 2, col = "blue", untf = T)

axis(side = 1, las = 1, at = c(seq(-3, 1, 1) + 2.5), 
     labels = expression(10^-3, 10^-2, 10^-1, 1, 10^1))
axis(side = 2, las = 1, at = c(seq(-5,3,2)), 
     labels = expression(10^-5, 10^-3, 10^-1, 10^1, 10^3)) 

legend("bottomright", legend = c("linear model", "isometric expectation", 
                                 "slope t-test: p = < 0.001"), 
       lty = c(2, 3, 0), col = c("red", "blue", "black"), 
       bty = "n", cex = 0.6)

mtext("Non-Ultrametic Branch Length\n(Substitution Rate)", side = 1,line = 4, cex = 1.25)
mtext("Ultrametic Branch Length\n(Time)", side = 2, line = 3, cex = 1.25)

box(lwd = 2)

# Close Plot Device
dev.off()
graphics.off()

rm(mod1)
```
## Show Plot
```{r}
img <- readPNG("../figures/Branch_Comparision.png")
grid.raster(img)
```

## Phylogeny Topological Features
```{r}
# This takes awhile, so you can turn it off

if (run.all == TRUE){
  tree.thres <- data.frame(matrix(NA, 100, 5))
  colnames(tree.thres) <- c("threshold", "nodes", "polytomies", "avg.polytomy")
  tree.thres$threshold <- seq(0.01, 0.15, length.out = 100)
  tree.thres$threshold <- 10^c(seq(-3, 0, length.out = 100))
  
  um.thres <- data.frame(matrix(NA, 100, 5))
  colnames(um.thres) <- c("threshold", "nodes", "polytomies", "avg.polytomy")
  um.thres$threshold <- 10^c(seq(-3, 2, length.out = 100))
  
  phy1 <- livingTree.r2
  phy2 <- livingTree.um
  
  for (i in 1:dim(tree.thres)[1]){
    phy.grained <- di2multi2(phy1, tree.thres[i,1])
    polys <- table(table(phy.grained$edge[,1]))
    poly.size <- table(phy.grained$edge[,1])
    tree.thres[i, 2] <- phy.grained$Nnode
    tree.thres[i, 3] <- sum(polys[which(as.numeric(names(polys)) > 2)]) 
    tree.thres[i, 4] <- mean(poly.size[which(poly.size > 2)])
    tree.thres[i, 5] <- length(poly.size[which(poly.size > 3)])
  }
  
  for (i in 1:dim(um.thres)[1]){
    phy.grained <- di2multi2(phy2, um.thres[i,1])
    polys <- table(table(phy.grained$edge[,1]))
    poly.size <- table(phy.grained$edge[,1])
    um.thres[i, 2] <- phy.grained$Nnode
    um.thres[i, 3] <- sum(polys[which(as.numeric(names(polys)) > 2)]) 
    um.thres[i, 4] <- mean(poly.size[which(poly.size > 2)])
    um.thres[i, 5] <- length(poly.size[which(poly.size > 3)])
  }
}
```

```{r, results = "hide"}
if (run.all == TRUE){
  png(filename="../figures/Tree_Topology.png",
      width = 1600, height = 1600, res = 96*2)
  
  layout(matrix(1:6, 3))
  par(mar = c(2,4,0.5,0.5), oma = c(3,1.5,3.5,1.5))
  plot(tree.thres$nodes ~ log10(tree.thres$threshold),
       xlab = "", ylab = "", type = "l", lwd = 2,
       xaxt = "n", ylim = c(0,15000))
  axis(side = 1, at = seq(-3,0,1), labels = c(expression(10^-3, 10^-2, 10^-1, 1)))
  mtext("# Nodes", side = 2, line = 2.5)
  mtext("Original Tree", side = 3, line = 1, outer = F)
  box(lwd = 2)
  plot(tree.thres$polytomies ~ log10(tree.thres$threshold),
       xlab = "", ylab = "", type = "l", lwd = 2, 
       xaxt = "n", ylim = c(0,2000))
  axis(side = 1, at = seq(-3,0,1), labels = c(expression(10^-3, 10^-2, 10^-1, 1)))
  mtext("# Polytomies", side = 2, line = 2.5)
  box(lwd = 2)
  plot(log10(tree.thres$avg.polytomy) ~ log10(tree.thres$threshold),
       xlab = "", ylab = "", type = "l", lwd = 2, 
       xaxt = "n", ylim = c(0,2.5))
  axis(side = 1, at = seq(-3,0,1), labels = c(expression(10^-3, 10^-2, 10^-1, 1)))
  mtext("Threshold (substitutions)", side =1, line = 2.5, cex = 1)
  mtext("Average Polytomy Size", side = 2, line = 2.5)
  box(lwd = 2)
  
  plot(um.thres$nodes ~ log10(um.thres$threshold),
       xlab = "", ylab = "", type = "l", lwd = 2,
       xaxt = "n", ylim = c(0,15000), xlim = c(-3,2))
  axis(side = 1, at = seq(-3,2,1), 
       labels = c(expression(10^-3, 10^-2, 10^-1, 1, 10^1, 10^2)))
  mtext("# Nodes", side = 2, line = 2.5)
  mtext("Ultrametric Tree", side = 3, line = 1, outer = F)
  box(lwd = 2)
  plot(um.thres$polytomies ~ log10(um.thres$threshold),
       xlab = "", ylab = "", type = "l", lwd = 2,
       xaxt = "n", ylim = c(0,1500), xlim = c(-3,2))
  axis(side = 1, at = seq(-3,2,1), 
       labels = c(expression(10^-3, 10^-2, 10^-1, 1, 10^1, 10^2)))
  mtext("# Polytomies", side = 2, line = 2.5)
  box(lwd = 2)
  plot(um.thres$avg.polytomy ~ log10(um.thres$threshold),
       xlab = "", ylab = "", type = "l", lwd = 2,
       xaxt = "n", ylim = c(0,50), xlim = c(-3,2))
  axis(side = 1, at = seq(-3,2,1), 
       labels = c(expression(10^-3, 10^-2, 10^-1, 1, 10^1, 10^2)))
  mtext("Threshold (Ma)", side =1, line = 2.5, cex = 1)
  mtext("Average Polytomy", side = 2, line = 2.5)
  box(lwd = 2)
  
  
  # Close Plot Device
  dev.off()
  graphics.off()
  
  # Clean Up
}
```
## Show Plot
```{r}
img <- readPNG("../figures/Tree_Topology.png")
grid.raster(img)
```

## Calculate Local Minimums from Edge Lengths and Maximum from Polytomies
```{r}
den.dis <- density(log10(LTP.edges))
mins <- peaks(den.dis$x, den.dis$y, mode = "min")
maxs <- peaks(den.dis$x, den.dis$y, mode = "max")
```

## Obsever Polytomies through time
```{r}
if (run.all == TRUE){
  time.thres <- data.frame(matrix(NA, 100, 4))
  colnames(time.thres) <- c("threshold", "Nnodes", "Npolytomies", "avg.polytomy")
  time.thres$threshold <- 10^c(seq(-3, 0, length.out = 100))
  
  thres.info <- vector(mode = "list", length = 100)
  
  phy1 <- livingTree.r2
  
  for (i in 1:dim(time.thres)[1]){
    phy.grained <- di2multi2(phy1, time.thres[i,1])
    poly.size <- table(phy.grained$edge[,1])
    polys <- table(table(phy.grained$edge[,1]))
    time.thres[i, 2] <- phy.grained$Nnode
    time.thres[i, 3] <- length(poly.size[which(poly.size > 2)])
    time.thres[i, 4] <- mean(poly.size[which(poly.size > 2)])
    aged <- node.age(phy.grained)
    node.info <- data.frame(matrix(NA, 4, length(aged$node.label) - 1))
    row.names(node.info) <- c("Nodes", "Node_Age", "Downstream_Branches",
                                               "Length_Upstream")
    node.info[1, ] <- as.numeric(aged$node.label[2:length(aged$node.label)])
    tips <- length(aged$tip.label)
    nodes <- aged$Nnode
    node.info[2, ] <- as.numeric(aged$ages[(tips + 2):(tips + nodes)])
    for (j in 2:length(aged$node.label)){
      node.info[3, j - 1] <- as.numeric(length(which(aged$edge[, 1] == (tips + j))))
      node.info[4, j - 1] <- as.numeric(aged$edge.length[which(aged$edge[ , 2] == (tips + j))])
    }
    thres.info[[i]] <- as.data.frame(t(as.matrix((node.info))))
  }
  
  # Explore Node Ages with Threshold
  # i = 80
  # test <- as.data.frame(thres.info[[i]])
  # plot(Node_Age ~ log10(Length_Upstream), 
  #      data = test[which(test$Downstream_Branches > 2), ])
  # legend("bottomleft",legend = paste("threshold =",
  #        round(log10(time.thres$threshold[i]), 5)), 
  #        bg = "gray", text.width = 0.5, box.col = "gray")
}


```


# Identify polytomies in gene tree and plot in the ultrametric tree
```{r}
# Trees
str(livingTree.r2)
str(livingTree.um)

LTP.r.edges <- livingTree.r2$edge.length
LTP.u.edges <- livingTree.um$edge.length
den.dis.r <- density(log10(LTP.r.edges))
den.dis.u <- density(log10(LTP.u.edges))
mins.r <- peaks(den.dis.r$x, den.dis.r$y, mode = "min")
mins.u <- peaks(den.dis.u$x, den.dis.u$y, mode = "min")

mins.r$x
mins.u$x
thres.test1 <- mins.r$x[1]
thres.test2 <- mins.r$x[2]


test.phy <- di2multi3(phy1 = livingTree.r2, 
                      phy2 = livingTree.um, tol = (10^thres.test1) * 10)

poly.node <- test.phy$multi.new
test.phy <- groupOTU(test.phy, groupInfo, group_name = "Domain")
levels(attributes(test.phy)$Domain) <- c("root", names(groupInfo), "polytomy")

test.tree <- ggtree(test.phy, aes(color = Domain), layout="rectangular")
test.tree + geom_point2(aes(subset=(node == 11934)), size=1, 
                         shape=18, colour="black", fill = "gray80") + 
  geom_point2(aes(subset=(node %in% c(poly.node))), size = 3, shape=18,
              colour=ggplotColours(n = 4)[1], fill = "gray80") + 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
  scale_color_manual(name = "", values=c(ggplotColours(n = 4)[1:3]), 
                     labels = c("polytomy", "Archaea", "Bacteria"))
ggsave("../figures/Test_Tree.png", width = 6, height = 4, dpi = 250)
rm(test.tree)

```

# Polytomy Movie
```{r}
if (run.all == TRUE){
  system("rm ../figures/movie/*.png")
  
  threshold <- 10^c(seq(-2.5, -1, length.out = 50))
  
  for (i in threshold){
    thres = i
    outname <- paste("../figures/movie/T", round(thres, 4), ".png", sep = "")
    test.phy <- di2multi3(phy1 = livingTree.r2, phy2 = livingTree.um, 
                          tol = thres)
    poly.node <- test.phy$multi.new
    test.phy <- node.age(test.phy)
    test.phy <- groupOTU(test.phy, groupInfo, group_name = "Domain")
    levels(attributes(test.phy)$Domain) <- c("polytomy", names(groupInfo))
    
    test.tree <- ggtree(test.phy, aes(color = Domain), layout="rectangular") 
    test.tree + geom_point2(aes(subset=(node %in% c(poly.node))), size = 3, shape=18,
                  colour=ggplotColours(n = 4)[1], fill = "gray80") + 
      theme_tree2(legend.position="top", legend.key = element_rect(colour = NA)) + 
      scale_color_manual(name = "", values=c(ggplotColours(n = 4)[c(1:3, 1)]), 
                         labels = c("polytomy", "Archaea", "Bacteria")) + 
      annotate("text", hjust = 0, x = max(test.phy$ages) * 0.05, 
               y = length(test.phy$tip.label) * 0.95, 
               label = paste("Threshold =", round(thres, 4)))
    ggsave(outname, width = 6, height = 4, dpi = 250)
    print(paste("Threshold", round(i, 4), "done"))
  }
  
  rm(test.tree)
  
  system("convert -loop 10 -delay 100 
         '../figures/movie/T*.png' '../figures/movie/Test.gif'")
}
```

# Test Geological Window
```{r}
thres = 0.01

test.phy <- di2multi3(phy1 = livingTree.r2, phy2 = livingTree.um, 
                        tol = thres)
poly.node <- test.phy$multi.new
test.phy <- node.age(test.phy)
test.phy <- groupOTU(test.phy, groupInfo, group_name = "Domain")
levels(attributes(test.phy)$Domain) <- c("polytomy", names(groupInfo))
  
geo.test <- data.frame(xmin = 3200, xmax = 3600, ymin = -Inf, ymax = Inf)

test.tree <- ggtree(test.phy, aes(color = Domain), layout="rectangular")

test.tree + geom_point2(aes(subset=(node %in% c(poly.node))), size = 3, 
                        shape=18, colour=ggplotColours(n = 4)[1], 
                        fill = "gray80") + 
  theme_tree2(legend.position="top", legend.key = element_rect(colour = NA)) + 
  scale_color_manual(name = "", values=c(ggplotColours(n = 4)[c(1:3, 1)]), 
                     labels = c("polytomy", "Archaea", "Bacteria")) + 
  annotate("text", hjust = 0, x = max(test.phy$ages) * 0.05, 
           y = length(test.phy$tip.label) * 0.95, 
           label = paste("Threshold =", round(thres, 4))) + 
  geom_rect(data = geo.test, aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax),
              color="grey80", fill = "gray", alpha=0.5, inherit.aes = FALSE)

rm(test.tree)
```

# Import Trait Data
```{r}
traits <- read.delim("../data/LTPs123.faprotax") , header = T, row.names = 1)
LTP.tax <- read.tax("../data/LTPs123_unique.nr_v123.wang.taxonomy", format = "rdp", col.tax = 2)
LTP.tax2 <- LTP.tax[match(livingTree.um$tip.label, LTP.tax$OTU), ]
LTP.tax <- read.delim("../data/LTPs123_unique.full.taxonomy")
LTP.tax2 <- LTP.tax[match(livingTree.um$tip.label, LTP.tax$OTU), ]

traits[1:5, 1:5]
colSums(traits)

sum(colSums(traits) == 0)
sum(rowSums(traits) == 0)

unknowns <- LTP.tax2[which(LTP.tax2$OTU %in% names(which(colSums(traits) == 0))), ]
dim(unknowns)
unknowns[1:5, ]

str(test.phy)
is.ultrametric(test.phy)
test.phy2 <- drop.tip(test.phy, c(unknowns$OTU))

traits2 <- traits[, which(!colnames(traits) %in% unknowns$OTU)]


c1.metab <- as.data.frame(t(traits2[1:10, ]))
#c1.metab <- c1.metab[, c(which(colSums(c1.metab) > 0))]

chemo.metab <- as.data.frame(as.numeric(t(traits[90,])))

```


# Add Traits to Tree
```{r}
geo.test <- data.frame(xmin = 2500, xmax = 4000, ymin = -Inf, ymax = Inf)

trait <- chemo.metab
colnames(trait) <- "Chemo.Het"

thres = 0.01

str(test.phy)
is.ultrametric(test.phy)

# Remove Tips with Unknown Traits
unkns <- as.character(unknowns$OTU)
length(unkns)
length(livingTree.um$tip.label)
sum(livingTree.um$tip.label %in% unkns)
sub.Tree.um <- drop.tip(livingTree.um, unkns, trim.internal = T)
sub.Tree.org <- drop.tip(livingTree.r2, unkns, trim.internal = T)

sub.Tree.multi <- di2multi3(phy1 = sub.Tree.org, phy2 = sub.Tree.um, 
                        tol = thres)

poly.node <- sub.Tree.multi$multi.new
sub.Tree.multi2 <- node.age(sub.Tree.multi)
sub.Tree.multi2 <- groupOTU(sub.Tree.multi2, groupInfo, group_name = "Domain")
levels(attributes(sub.Tree.multi2)$Domain) <- c("root", names(groupInfo))

# Plot Tree With Traits
test.tree <- ggtree(tr = sub.Tree.multi2, aes(color = Domain), layout="rectangular") + 
  geom_point2(aes(subset=(node %in% c(poly.node))), size = 3, shape=18,
              colour=ggplotColours(n = 4)[1], fill = "gray80") + 
  geom_rect(data = geo.test, aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax),
              color="grey", fill = "gray", alpha=0.25, inherit.aes = FALSE)

test.tree

c1.metab <- as.data.frame(t(traits2[1:10, ]))
#c1.metab <- c1.metab[, c(which(colSums(c1.metab) > 0))]

chemo.metab <- as.data.frame(as.numeric(t(traits2[90,])))
trait <- chemo.metab
colnames(trait) <- "Chemo.Het"
rownames(trait) <- colnames(traits2)

gheatmap2(test.tree, trait, high = "red", low = "white", color = "white", 
          offset = 1, width = 0.1, font.size = 2, colnames_position = "top") + 
theme_tree2(legend.position="top", legend.key = element_rect(colour = NA)) +
  scale_color_manual(name = "", values=c(ggplotColours(n = 4)[c(1:3, 1)]), 
                     labels = c("polytomy", "Archaea", "Bacteria")) + 
  annotate("text", hjust = 0, x = max(sub.Tree.multi2$ages) * 0.05, 
           y = length(sub.Tree.multi2$tip.label) * 0.95, 
           label = paste("Threshold =", round(thres, 4)))


test.tree <- ggtree(sub.Tree.multi2, aes(color = Domain), layout="rectangular") + 
  geom_rect(data = geo.test, aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax),
              color="grey80", fill = "gray", alpha=0.5, inherit.aes = FALSE)

geo.test <- data.frame(xmin = 1800, xmax = 2200, ymin = -Inf, ymax = Inf)

edge.col <- attributes(sub.Tree.multi2)$Domain


raw <- gheatmap2(test.tree, trait, high = "red", low = "white", color = NULL, 
          offset = 1, width = 0.1, font.size = 2,
          colnames_position = "top", guide = "none") + 
  geom_point2(aes(subset=(node %in% c(poly.node))), size = 3, shape=18, 
              colour=ggplotColours(n = 4)[1], fill = "gray80") + 
  theme_tree2(legend.position="top", legend.key = element_rect(colour = NA)) + 
  scale_alpha(guide = "none") + 
  scale_color_manual(name = "", values=c(ggplotColours(n = 4)[c(1:3, 1)]), 
                     labels = c("polytomy", "Archaea", "Bacteria")) + 
  annotate("text", hjust = 0, x = max(sub.Tree.multi2$ages) * 0.05, 
           y = length(sub.Tree.multi2$tip.label) * 0.95, 
           label = paste("Threshold =", round(thres, 4))) 
  scale_x_ggtree(raw)

# + 
  #geom_rect(data = geo.test, aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax),
  #            color="grey80", fill = "gray", alpha=0.5, inherit.aes = FALSE)

par(oma = c(4, 1, 1, 1))
edge.col <- attributes(sub.Tree.multi2)$Domain
main <- plot(sub.Tree.multi2, edge.color = edge.col, show.tip.label = FALSE,
             no.margin = T) # , y.lim = c(length(sub.Tree.multi2$tip.label), 0))
axis(side = 1, at = c(seq(0, 5000, 1000)))

str(main)

trait.test <- edge.col

```

# Geological Map 
```{r}
# Define Geological Events
geol <- data.frame(matrix(NA, 3, 4))
colnames(geol) <- c("event", "date", "min", "max")
geol[1, ] <- c("Great Oxidation Event", 2300, 1800, 2500)
geol[2, ] <- c("Methanian", 2700, 2500, 2800)
geol[3, ] <- c("Sulfur Dominance", 3000, 2800, 3500)

geol$date <- as.numeric(geol$date)
geol$min <- as.numeric(geol$min)
geol$max <- as.numeric(geol$max)

geol.r <- 4500 - geol[, 2:4]

geo.ox <- data.frame(xmin = geol.r[1, 2], xmax = geol.r[1, 3], 
                     ymin = -Inf, ymax = Inf)
geo.met <- data.frame(xmin = geol.r[2, 2], xmax = geol.r[2, 3], 
                      ymin = -Inf, ymax = Inf)
geo.sul <- data.frame(xmin = geol.r[3, 2], xmax = geol.r[3, 3], 
                      ymin = -Inf, ymax = Inf)


# Define Traits
rownames(traits2)
rownames(traits2)[grep("*cyano*", rownames(traits2))]

chem <- colSums(traits2[90, ])
phot <- colSums(traits2[grep("*photo*", rownames(traits2)), ]) >= 1
pho2 <- colSums(traits2[grep("*anoxygenic_photo*", rownames(traits2)), ]) >= 1
meth <- colSums(traits2[grep("*meth*", rownames(traits2)), ]) >= 1
sulf <- colSums(traits2[grep("*sul*", rownames(traits2)), ]) >= 1
deni <- colSums(traits2[grep("*denitri*", rownames(traits2)), ]) >= 1
trait <- data.frame(chem, phot, meth, sulf, deni) 

thres = 0.075
sub.Tree.multi <- di2multi3(phy1 = sub.Tree.org, phy2 = sub.Tree.um, 
                        tol = thres)

poly.node <- sub.Tree.multi$multi.new
sub.Tree.multi2 <- node.age(sub.Tree.multi)
sub.Tree.multi2 <- groupOTU(sub.Tree.multi2, groupInfo, group_name = "Domain")
levels(attributes(sub.Tree.multi2)$Domain) <- c("root", names(groupInfo))


geo.tree <- ggtree(sub.Tree.multi2, aes(color = Domain), layout="rectangular") + 
  geom_rect(data = geo.ox, aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax),
              color="grey85", fill = "gray85", inherit.aes = FALSE) + 
  geom_rect(data = geo.met, aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax),
              color="grey90", fill = "gray90", inherit.aes = FALSE) + 
  geom_rect(data = geo.sul, aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax),
              color="grey95", fill = "gray95", inherit.aes = FALSE)
  geo.tree$layers <- rev(geo.tree$layers)



geo.traits <- gheatmap2(geo.tree, trait, high = "red", low = "white", color = NULL, 
          offset = 2, width = 0.2, font.size = 2.5,
          colnames_position = "top", guide = "none") + 
  geom_point2(aes(subset=(node %in% c(poly.node))), size = 3, shape=18, 
              colour=ggplotColours(n = 4)[1], fill = "gray80") + 
  theme_tree2(legend.position="top", legend.key = element_rect(colour = NA),
              plot.margin = unit(c(1,1,1,1), "cm")) + 
  scale_alpha(guide = "none") + 
  scale_color_manual(name = "", values=c(ggplotColours(n = 4)[c(1:3, 1)]), 
                     labels = c("polytomy", "Archaea", "Bacteria")) + 
  annotate("text", hjust = 0, x = 0 , y = length(sub.Tree.multi2$tip.label), 
           label = paste("Threshold =", round(thres, 4)), size = 3) + 
  annotate("text", hjust = 0, x = geol.r[1,1] , y = length(sub.Tree.multi2$tip.label) * 0.6, 
           label = "Great Oxidation", size = 3, angle = 90, vjust = 1) +
  annotate("text", hjust = 0, x = geol.r[2,1] , y = length(sub.Tree.multi2$tip.label) * 0.6, 
           label = "Methane Era", size = 3, angle = 90, vjust = 1) +
  annotate("text", hjust = 0, x = geol.r[3,1] , y = length(sub.Tree.multi2$tip.label) * 0.6, 
           label = "Sulfur Era", size = 3, angle = 90, vjust = 1) +
  scale_x_continuous(breaks = seq(0,4000, 1000), expand = c(0.01,0))

png(filename="../figures/GeoTree.png", width = 2000, height = 800, res = 96*2)

geo.traits

dev.off()
graphics.off()

img <- readPNG("../figures/GeoTree.png")
grid.raster(img)

```


# Identify a specific trait
```{r}
str(traits2)
rownames(traits2)[grep("denit*", rownames(traits2))]
denit <- traits2[grep("denit*", rownames(traits2)), ]
denit <- denit[, which(colSums(denit) > 0)]
denit
LTP.tax2[which(LTP.tax2$OTU %in% colnames(denit)), 1:7]


LTP.tax2[grep("*denit*", LTP.tax2$Genus), ]




```




# Repeat with Other trees (still working here)

## NTL Exploratory Tree with Domain Grouping
```{r, results = "hide"}
png(filename="../figures/NTL_explore.png",
    width = 1600, height = 1600, res = 96*2)

NTL.2 <- NTL
NTL.tax2 <- NTL.tax[na.omit(match(NTL.2$tip.label, NTL.tax$OTU)), ]
NTL.tax2[which(NTL.tax2$Domain == "unknown"), ]$Domain <- "Bacteria"
groupInfo <- split(NTL.2$tip.label, NTL.tax2$Domain)
newTree <- groupOTU(NTL.2, groupInfo, group_name = "Domain")
levels(attributes(newTree)$Domain) <- c("root", names(groupInfo))
newTree$node.label[which(attributes(newTree)$Domain == "root")]
ggtree(newTree, aes(color = Domain), layout="rectangular") + 
  geom_text(aes(subset=!isTip, label = node), show.legend = FALSE,
            hjust=-0.1, vjust = -0.5, size = 2) + 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) 

# Close Plot Device
dev.off()
graphics.off()
```
## Show Plot
```{r}
img <- readPNG("../figures/NTL_explore.png")
grid.raster(img)
```

Instead of trying to remove the Eukaryota from this tree, I decided to go back
to the original data and remove the Eukaryote manually and then generate a new
tree and taxonomy file. I will import them here and continue. 

## Import Tree Files and Taxonomy Information
```{r}
NTL <- read.tree("../data/NTL.clean.tree")

NTL.tax <- read.tax("../data/NTL.clean.pds.wang.taxonomy", format = "rdp", col.tax = 2, tax.levels = 6)



NTL <- read.tree("../data/NTL.tree")

NTL.tax <- read.tax("../data/nmicrobiol201648_s7.pds.wang.taxonomy", format = "rdp", col.tax = 2, tax.levels = 6)




# Reorder Tax
NTL.tax <- NTL.tax[match(NTL$tip.label, NTL.tax$OTU), ]
all.equal(as.character(NTL$tip.label), as.character(NTL.tax$OTU))

quantile(NTL$edge.length)
#NTL$edge.length[which(NTL$edge.length == 0)] <- 0.0001
#quantile(NTL$edge.length)


```


# Plot Tree with Midpoint Root

```{r, results = "hide"}
png(filename="../figures/NTL_root.png",
    width = 1600, height = 1600, res = 96*2)

#domain <- sapply(strsplit(NTL$tip.label, "_"), "[", 1)
#groupInfo <- split(NTL$tip.label, domain)

sum(NTL$tip.label != NTL.tax$OTU)

groupInfo <- split(NTL$tip.label, NTL.tax$Domain)
NTL.g <- groupOTU(NTL, groupInfo, group_name = "Domain")

NTL.r <- midpoint.root(NTL.g)

NTL.tax2 <- NTL.tax[match(NTL.r$tip.label, NTL.tax$OTU), ]
NTL.tax2$Domain[which(NTL.tax2$Domain == "unknown")] <- "Bacteria"
groupInfo <- split(NTL.r$tip.label, NTL.tax2$Domain)
NTL.r2 <- groupOTU(NTL.r, groupInfo, group_name = "Domain")

levels(attributes(NTL.r2)$Domain) <- c("root", names(groupInfo))
#levels(attributes(NTL.g)$Domain) <- c(names(groupInfo))

rt <- ggtree(NTL.r2, aes(color = Domain), layout="rectangular")
rt + geom_point2(aes(subset=(node == 11934)), size=4, shape=18, 
                 colour=ggplotColours(n = 4)[1], fill = "gray80") + 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
  scale_color_manual(values=c(ggplotColours(n = 4)[1:4])) 

# Close Plot Device
dev.off()
graphics.off()

# Clean Up
rm(rt)
```
## Show Plot
```{r}
img <- readPNG("../figures/NTL_root.png")
grid.raster(img)
```


```{r}
NTL.3 <- drop.tip(NTL.r2, c(NTL.tax2$OTU[which(NTL.tax2$Domain == "Eukaryota")]))
NTL.tax.3 <- NTL.tax2[which(NTL.tax2$Domain != "Eukaryota"), ]

NTL.tax.3 <- NTL.tax.3[match(NTL.3$tip.label, NTL.tax.3$OTU), ]
length(NTL.3$tip.label)
length(NTL.tax.3$OTU)
sum(NTL.3$tip.label != NTL.tax.3$OTU)

groupInfo <- split(NTL.3$tip.label, NTL.tax.3$Domain)
NTL.3 <- groupOTU(NTL.3, groupInfo, group_name = "Domain")

levels(attributes(NTL.3)$Domain) <- c("root", names(groupInfo))
#levels(attributes(NTL.g)$Domain) <- c(names(groupInfo))

rt <- ggtree(NTL.3, aes(color = Domain), layout="rectangular")
rt + geom_point2(aes(subset=(node == 11934)), size=4, shape=18, 
                 colour=ggplotColours(n = 4)[1], fill = "gray80") + 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
  scale_color_manual(values=c(ggplotColours(n = 4)[1:4])) 

```


## Export Rooted Tree
```{r}
write.tree(NTL.3, file = "../data/NTL.rooted.tre")
write.nexus(NTL.3, file = "../data/NTL.rooted.nex")
```

## Identify the NTL Node that seperates bacteria and archaea
```{r, results = "hide"}
png(filename="../figures/NTL_archaea.png",
    width = 1600, height = 1600, res = 96*2)


groupInfo <- split(NTL.3$tip.label, NTL.tax.3$Domain)
NTL.3 <- groupOTU(NTL.3, groupInfo, group_name = "Domain")

par(mar=c(1,1,1,3))
cp <- ggtree(NTL.3, aes(color = Domain), layout="rectangular")
cp + geom_point2(aes(subset=(node == 18227)), size=3, shape=18, 
                 colour=ggplotColours(n = 4)[1], fill = "gray80") + 
 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
  geom_text2(aes(subset=!isTip, label = node), hjust=-0.1, vjust = -0.5, size = 2) + 
  guides(colour = guide_legend(override.aes = list(shape = NA, label = NA))) + 
  scale_color_manual(values=c(ggplotColours(n = 4)[1:3])) 

# Close Plot Device
dev.off()
graphics.off()

# Clean Up
rm(cp)
```
## Show Plot
```{r}
img <- readPNG("../figures/NTL_archaea.png")
grid.raster(img)
```

The Bac/Arc node is 1715

# ID Nodes that separate
```{r}
D <- cophenetic(NTL.3)
dd <- max(D)
ii <- which(D == dd)[1]
ii <- c(ceiling(ii/nrow(D)), ii%%nrow(D))
if (ii[2] == 0) 
  ii[2] <- nrow(D)
spp <- rownames(D)[ii]
nn <- which(NTL.3$tip.label == spp[2])



```

# Isolate Cyanobacteria and Identify the LTP Node that seperates cyanos
```{r, results = "hide"}
png(filename="../figures/NTL_Cyano.png",
    width = 1600, height = 1600, res = 96*2)

sum(NTL.tax.3$OTU != NTL.3$tip.label)

sum(NTL.tax.3$Class == "Cyanobacteria")

Domain.NTL <- NTL.tax.3$Domain
for (i in 1:length(Domain.NTL)){
  if (NTL.tax.3$Class[i] == "Cyanobacteria") {
    Domain.NTL[i] <- "Cyanobacteria"
  }
}

groupInfo <- split(NTL.3$tip.label, Domain.NTL)
NTL.3 <- groupOTU(NTL.3, groupInfo, group_name = "Domain")

cp <- ggtree(NTL.3, aes(color = Domain), layout="rectangular")
cp + 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
  geom_text2(aes(subset=!isTip, label = node), hjust=-0.1, vjust = -0.5, size = 2) + 
  guides(colour = guide_legend(override.aes = list(shape = NA, label = NA))) + 
  scale_color_manual(values=c(ggplotColours(n = 4)[1:4])) 

# Extract Clade 2018
# viewClade(cp + geom_tiplab() +
#             geom_text(aes(subset=!isTip, label = node), show.legend = FALSE,
#             hjust=-0.1, vjust = -0.5, size = 2) + 
#             theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
#             scale_color_manual(values=c(ggplotColours(n = 4)[c(3,4,2,1)])), 
#           node=2018)

# Close Plot Device
dev.off()
graphics.off()

# Clean Up
#```
## Show Plot
#```{r}
img <- readPNG("../figures/NTL_Cyano.png")
grid.raster(img)
```

# Geological Map
```{r}
# Define Geological Events
geol <- data.frame(matrix(NA, 3, 4))
colnames(geol) <- c("event", "date", "min", "max")
geol[1, ] <- c("Great Oxidation Event", 2300, 1800, 2500)
geol[2, ] <- c("Methanian", 2700, 2500, 2800)
geol[3, ] <- c("Sulfur Dominance", 3000, 2800, 3500)

geol$date <- as.numeric(geol$date)
geol$min <- as.numeric(geol$min)
geol$max <- as.numeric(geol$max)

geol.r <- 4500 - geol[, 2:4]

geo.ox <- data.frame(xmin = geol.r[1, 2], xmax = geol.r[1, 3], 
                     ymin = -Inf, ymax = Inf)
geo.met <- data.frame(xmin = geol.r[2, 2], xmax = geol.r[2, 3], 
                      ymin = -Inf, ymax = Inf)
geo.sul <- data.frame(xmin = geol.r[3, 2], xmax = geol.r[3, 3], 
                      ymin = -Inf, ymax = Inf)

NTL.um <- read.tree("../data/NTL.dated.tre")

png(filename="../figures/NTL_dated.png",
    width = 1600, height = 1600, res = 96*2)

sum(NTL.um$tip.label, NTL.tax.3$OTU)
groupInfo <- split(NTL.3$tip.label, NTL.tax.3$Domain)
NTL.3 <- groupOTU(NTL.3, groupInfo, group_name = "Domain")

NTL.um <- groupOTU(NTL.um, groupInfo, group_name = "Domain")
levels(attributes(NTL.um)$Domain) <- c("root", names(groupInfo))
um <- ggtree(NTL.um, aes(color = Domain), layout="rectangular")
um + geom_point2(aes(subset=(node == 1715)), size=4, shape=18, 
                 colour=ggplotColours(n = 4)[1], fill = "gray80") + 
  theme(legend.position="top", legend.key = element_rect(colour = NA)) + 
  scale_color_manual(values=c(ggplotColours(n = 4)[1:3])) 

# Close Plot Device
dev.off()
graphics.off()

# Clear Up
rm(um)
```
## Show Plot
```{r}
img <- readPNG("../figures/NTL_dated.png")
grid.raster(img)
```

```{r}
NTLtraits <- t(read.delim("../data/NTL.faprotax", header = T, row.names = 1))

colSums(NTLtraits)
length(NTL.3$tip.label)

source("../bin/ConsenTrait.R")
ConsenTrait(NTL.3, NTLtraits[, 1:5], 0.9, "../data/ConsenTrait.test.txt")


```