---
title: "Microbial Innovations: IMG Gene Innovations"
author: "Mario E. Muscarella, James P. O'Dwyer"
date: "Last updated on `r format(Sys.time(), '%d %B, %Y')`"
header-includes:
  - \usepackage{array}
  - \usepackage{graphics}
output: 
  pdf_document: 
    fig_caption: yes
---


# Initial Setup
```{r}
# Clear Environment and Set Working Directory
rm(list=ls())
setwd("~/GitHub/microbial-innovations/analyses")

# Add Basic Functions
se <- function(x, ...){sd(x, na.rm = TRUE)/sqrt(length(na.omit(x)))}
ci <- function(x, ...){1.96 * sd(x,na.rm = TRUE)}

# Load Packages
require("ape"); require("png"); require("grid")
require("ggtree")
require("phytools")
library("colorspace")
require("simecol")
require("picante")
require("tidyr")

# Load Source Functions
source("../bin/MothurTools.R")
source("../bin/di2multi2.R")
source("../bin/TraitMap.R")
source("../bin/Scatterhist.R")

# Define Simple Functions
ggplotColours <- function(n = 6, h = c(0, 360) + 15){
  if ((diff(h) %% 360) < 1) h[2] <- h[2] - 360/n
  hcl(h = (seq(h[1], h[2], length = n)), c = 100, l = 65)
}

ttest <- function(reg, coefnum, val){
  co <- coef(summary(reg))
  tstat <- (co[coefnum,1]-val)/co[coefnum,2]
  pstat <- 2 * pt(abs(tstat), reg$df.residual, lower.tail = FALSE)
  return(list = c(t = tstat, df = reg$df.residual, p =  pstat))
}

# Save Default Plot Settings
opar <- par(no.readonly = TRUE)  # Saves plot defaults

# Run All?
run.all <- FALSE
```

# Import Tree, Taxonomy, and Traits 
```{r}
IMG.tree <- read.tree("../data/IMG.dated.tree")
IMG.meta <- read.delim("../data/taxontable.txt")
IMG.traits <- read.delim("../data/IMG.KEGG.trimmed.txt")

# Pull Out Taxonomy
# Removing Any Taxonomy not Matching
IMG.tax <- IMG.meta[which(IMG.meta$IMG.Genome.ID %in% IMG.tree$tip.label), ]
# Reorder to Match Tips
IMG.tax <- IMG.tax[order(match(as.character(IMG.tax$taxon_oid),
                               IMG.tree$tip.label)), ]

# Organize Traits
IMG.traits <- IMG.traits[order(match(as.character(IMG.traits$OID), 
                                     IMG.tree$tip.label)), ]

# Check That All Match
all.equal(IMG.tree$tip.label, as.character(IMG.tax$taxon_oid))
all.equal(IMG.tree$tip.label, as.character(IMG.traits$OID))

```

# Import KEGG Databases
```{r}
# Import KEGG Database
gene.database.temp <- read.table("../data/ko00001.keg", header = T, 
                                 fill = T, comment.char = "#", 
                                 quote = "", sep = "\t")

# Organize KEGG
A.level.index <- grep("A<b>", gene.database.temp$X.D)
A.levels <- gsub("A<b>", "", 
                 gsub("</b>", "",
                      as.character(gene.database.temp$X.D[A.level.index])))

B.level.index <- grep("B  <b>", gene.database.temp$X.D)
B.levels <- gsub("B  <b>", "", 
                 gsub("</b>", "",
                      as.character(gene.database.temp$X.D[B.level.index])))

C.level.index <- grep("C    ", gene.database.temp$X.D)
C.levels <- gsub("C    ", "", 
                 gsub("</b>", "",
                      as.character(gene.database.temp$X.D[C.level.index]))) 

D.level.index <- grep("D      K", gene.database.temp$X.D)
D.levels <- gsub("D      K", "K", 
                 as.character(gene.database.temp$X.D[D.level.index]))
D.ko <- sapply(strsplit(D.levels , "  "), function(x) x[1])
D.gene <- sapply(strsplit(D.levels , "  "), function(x) x[2])

# Create Gene Dictionary
gene.table <- data.frame(matrix(NA, nrow = length(D.levels), ncol = 5))
colnames(gene.table) <- c("A", "B", "C", "ko", "gene")
gene.table$ko <- D.ko
gene.table$gene <- D.gene

A.ranges <- c(A.level.index, max(D.level.index))
B.ranges <- c(B.level.index, max(D.level.index))
C.ranges <- c(C.level.index, max(D.level.index))

for(a in 1:length(A.level.index)){
  temp <- D.level.index[which(as.numeric(D.level.index) 
                              %in% seq(A.ranges[a], A.ranges[a+1]))]
  gene.table$A[D.level.index %in% temp] <- A.levels[a]
}

for(b in 1:length(B.level.index)){
  temp <- D.level.index[which(as.numeric(D.level.index) 
                              %in% seq(B.ranges[b], B.ranges[b+1]))]
  gene.table$B[D.level.index %in% temp] <- B.levels[b]
}

for(c in 1:length(C.level.index)){
  temp <- D.level.index[which(as.numeric(D.level.index) 
                              %in% seq(C.ranges[c], C.ranges[c+1]))]
  gene.table$C[D.level.index %in% temp] <- C.levels[c]
}
  
dim(gene.table)
head(gene.table)[1:5]

# Remove Human Disease
gene.dictionary <- gene.table[-c(which(gene.table$A == "Human Diseases")), ]

length(gene.dictionary$ko)
length(unique(gene.dictionary$ko))
```

# Trait Predictions
## ConsenTrait Prediction
```{r}
IMG.Cons <- read.delim("../data/IMG_ConsenTrait.txt")

# Create Gene Table
Gene.Cons <- as.data.frame(matrix(NA, ncol = 5, 
                           nrow = length(unique(IMG.Cons$trait))))
colnames(Gene.Cons) <- c("gene", "origins", "trait.mean", "trait.se", "nodes") 

# Node Correction (Con starts at 1 not at Ntips + 1)
Ntips <- length(IMG.tree$tip.label)
#IMG.Cons$node2 <- as.numeric(IMG.Cons$node) + Ntips

# Populate Gene Table
Gene.Cons$gene <- unique(IMG.Cons$trait)
for(i in 1:dim(Gene.Cons)[1]){
  gene <- Gene.Cons$gene[i]
  temp <- IMG.Cons[which(IMG.Cons$trait == gene), ]
  temp <- temp[which(temp$cluster_size > 1), ]
  nodes <- toString(temp$node, collapse = ", ")
  Gene.Cons$origins[i] <- dim(temp)[1]
  Gene.Cons$trait.mean[i] <- mean(temp$distance)
  Gene.Cons$trait.se[i] <- se(temp$distance)
  Gene.Cons$nodes[i] <- nodes
}
```

## ASR Predictions
```{r}
IMG.ASR.full <- read.delim("../data/IMG_ASR_Long.txt")

# Fetch Names from Trait Database
IMG.ASR.full$Name <- IMG.ASR.full$Trait

# Crate MYA time
IMG.ASR.full$Dist <- 44000 - IMG.ASR.full$R.Dist

# Create Gene Table
Gene.ASR.full <- as.data.frame(matrix(NA, ncol = 4, 
                           nrow = length(unique(IMG.ASR.full$Name))))
colnames(Gene.ASR.full) <- c("gene", "origins", "trait.mean", "trait.se")

# Populate Gene Table
Gene.ASR.full$gene <- unique(IMG.ASR.full$Name)
for(i in 1:dim(Gene.ASR.full)[1]){
  gene <- Gene.ASR.full$gene[i]
  temp <- IMG.ASR.full[which(IMG.ASR.full$Name == gene), ]
  Gene.ASR.full$origins[i] <- dim(temp)[1]
  Gene.ASR.full$trait.mean[i] <- mean(temp$Dist)
  Gene.ASR.full$trait.se[i] <- se(temp$Dist)
}

# Only ASR Origins
IMG.ASR <- read.delim("../data/IMG_genes_ASR.txt")
IMG.ASR$Name <- IMG.ASR$trait

Gene.ASR <- as.data.frame(matrix(NA, ncol = 7, 
                           nrow = length(unique(IMG.ASR$Name))))
colnames(Gene.ASR) <- c("gene", "x", "y", "origins", "trait.mean", 
                        "trait.se", "nodes")

Gene.ASR$gene <- as.character(sort(unique(IMG.ASR$Name)))
for(i in 1:dim(Gene.ASR)[1]){
  gene <- Gene.ASR$gene[i]
  temp <- IMG.ASR[which(IMG.ASR$Name == gene), ]
  temp.dates <- as.character(temp$Dates.Origins)
  temp.num <- as.numeric(unlist(strsplit(temp.dates, ",", fixed = T)))
  temp.num <- 4400 - temp.num
  Gene.ASR$x[i] <- temp$Rate1
  Gene.ASR$y[i] <- temp$Rate2
  Gene.ASR$origins[i] <- length(as.numeric(temp.num))
  Gene.ASR$trait.mean[i] <- mean(temp.num)
  Gene.ASR$trait.se[i] <- se(temp.num)
  Gene.ASR$nodes[i] <- as.character(temp$Origins)
}
```

## Origins and Rates
```{r}
Gene.ASR2 <- Gene.ASR[which(Gene.ASR$gene %in% gene.dictionary$ko), ]
Gene.Cons2 <- Gene.Cons[which(Gene.Cons$gene %in% gene.dictionary$ko), ]

gene.set <- union(Gene.ASR2$gene, Gene.Cons2$gene)

Gene.Origins <- as.data.frame(matrix(NA, nrow = length(gene.set),
                                  ncol = 10))
colnames(Gene.Origins) <- c("Gene", "Name", "X", "Y", 
                         "ASR.Mean", "ASR.SE", "ASR.origins",
                         "CON.Mean", "CON.SE", "CON.origins")

Gene.Origins$Gene <- sort(gene.set)

for(i in 1:dim(Gene.Origins)[1]){
  gene.temp <- Gene.Origins$Gene[i]
  asr.temp <- Gene.ASR2[which(Gene.ASR2$gene == gene.temp), ]
  con.temp <- Gene.Cons2[which(Gene.Cons2$gene == gene.temp), ]
  name.temp <- gene.dictionary$gene[which(gene.dictionary$ko ==
                                          Gene.Origins$Gene[i])]
  name.temp2 <- strsplit(name.temp, ";")[[1]][1]
  Gene.Origins$Name[i] <- name.temp2
  if(dim(asr.temp)[1] != 0){
  Gene.Origins$X[i] <- asr.temp$x
  Gene.Origins$Y[i] <- asr.temp$y
  Gene.Origins$ASR.Mean[i] <- asr.temp$trait.mean
  Gene.Origins$ASR.SE[i] <- asr.temp$trait.se
  Gene.Origins$ASR.origins[i] <- asr.temp$nodes
  }
  if(dim(con.temp)[1] != 0){
  Gene.Origins$CON.Mean[i] <- con.temp$trait.mean
  Gene.Origins$CON.SE[i] <- con.temp$trait.se
  Gene.Origins$CON.origins[i] <- con.temp$nodes
  }
}
```

## Origin Comparison
```{r}
png(filename="../figures/IMG_Origins.png",
    width = 1000, height = 1000, res = 96*2)

x = Gene.Origins$ASR.Mean
y = Gene.Origins$CON.Mean
zones=matrix(c(2,0,1,3), ncol=2, byrow=TRUE)
par(oma = c(1,1,1,1))
layout(zones, widths=c(4/5,1/5), heights=c(1/5,4/5))
xhist = hist(x, plot=FALSE, breaks = seq(0, 4400, 100))
yhist = hist(y, plot=FALSE, breaks = xhist$breaks)
top = max(c(xhist$counts, yhist$counts))
par(mar=c(5,6,1,1))
plot(x,y, las = 1, xlab = "", ylab = "",
     xlim = c(0, 4400), ylim = c(0, 4400),
     bg=rgb(0, 0, 1, 0.5), pch=21, axes = F)
axis(side = 1, at = c(seq(0, 4000, 1000)), las = 1, tck = -0.03)
axis(side = 2, at = c(seq(0, 4000, 1000)), las = 1, tck = -0.03)
axis(side = 3, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = -0.01)
axis(side = 4, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = -0.01)
axis(side = 1, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
axis(side = 2, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
axis(side = 3, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
axis(side = 4, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
box(lwd = 1.5)
mtext("ASR Origin (mya)", side=1, line=3, outer=F, cex = 1.5)
mtext("ConsenTrait Origin (mya)", side=2, line=4, outer=F, cex = 1.5)
par(mar=c(0,6,1,1))
barplot(xhist$counts[0:44], axes=FALSE, space=0, xlim = c(0, 44))
par(mar=c(5,0,1,1))
barplot(yhist$counts, axes=FALSE, space=0, horiz=TRUE)

# Close Plot Device
dev.off()
graphics.off()
#```
## Show Plot
#```{r}
img <- readPNG("../figures/IMG_Origins.png")
grid.raster(img)
```


## Multi Pathway Genes
```{r}
png(filename="../figures/IMG_NumOrigins.png",
    width = 1800, height = 900, res = 96*2)

multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))

Single.Gene.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% single), ]
x = Gene.Origins$ASR.Mean

zones=matrix(c(2,0,5,0,1,3,4,6), ncol=4, byrow=TRUE)
par(oma = c(6,6,3,1))
layout(zones, widths=c(4/5,1/5,4/5,1/5), heights=c(1/5,4/5))

# Single Pathway
x = Gene.Origins$ASR.Mean[which(Gene.Origins$Gene %in% single)]
y = Gene.Origins$CON.Mean[which(Gene.Origins$Gene %in% single)]

xhist = hist(x, plot=FALSE, breaks = seq(0, 4400, 100))
yhist = hist(y, plot=FALSE, breaks = xhist$breaks)
top = max(c(xhist$counts, yhist$counts))
par(mar=c(1,2,1,1))
plot(x,y, las = 1, xlab = "", ylab = "",
     xlim = c(0, 4400), ylim = c(0, 4400),
     bg=rgb(0, 0, 1, 0.5), pch=21, axes = F)
axis(side = 1, at = c(seq(0, 4000, 1000)), las = 1, tck = -0.03)
axis(side = 2, at = c(seq(0, 4000, 1000)), las = 1, tck = -0.03)
axis(side = 3, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = -0.01)
axis(side = 4, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = -0.01)
axis(side = 1, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
axis(side = 2, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
axis(side = 3, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
axis(side = 4, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
box(lwd = 1.5)
mtext("ASR Origin (mya)", side=1, line=3, outer=T, cex = 1.5)
mtext("ConsenTrait Origin (mya)", side=2, line=4, outer=F, cex = 1.5)
par(mar=c(0,2,1,1))
barplot(xhist$counts[0:44], axes=FALSE, space=0, xlim = c(0, 44))
mtext(side = 3, "Single Pathway Gene", line = 1, cex = 1.5)
par(mar=c(1,0,1,1))
barplot(yhist$counts, axes=FALSE, space=0, horiz=TRUE)

# Multiple Pathway
x = Gene.Origins$ASR.Mean[which(Gene.Origins$Gene %in% multi)]
y = Gene.Origins$CON.Mean[which(Gene.Origins$Gene %in% multi)]

xhist = hist(x, plot=FALSE, breaks = seq(0, 4400, 100))
yhist = hist(y, plot=FALSE, breaks = xhist$breaks)
top = max(c(xhist$counts, yhist$counts))
par(mar=c(1,2,1,1))
plot(x,y, las = 1, xlab = "", ylab = "",
     xlim = c(0, 4400), ylim = c(0, 4400),
     bg=rgb(0, 0, 1, 0.5), pch=21, axes = F)
axis(side = 1, at = c(seq(0, 4000, 1000)), las = 1, tck = -0.03)
axis(side = 2, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = -0.03)
axis(side = 3, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = -0.01)
axis(side = 4, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = -0.01)
axis(side = 1, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
axis(side = 2, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
axis(side = 3, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
axis(side = 4, at = c(seq(0, 4000, 1000)), labels = F, las = 1, tck = 0.01)
box(lwd = 1.5)
#mtext("ASR Origin (mya)", side=1, line=3, outer=F, cex = 1.5)
#mtext("ConsenTrait Origin (mya)", side=2, line=4, outer=F, cex = 1.5)
par(mar=c(0,2,1,1))
barplot(xhist$counts[0:44], axes=FALSE, space=0, xlim = c(0, 44))
mtext(side = 3, "Multi Pathway Gene", line = 1, cex = 1.5)
par(mar=c(1,0,1,1))
barplot(yhist$counts, axes=FALSE, space=0, horiz=TRUE)

# Close Plot Device
dev.off()
graphics.off()
#```
## Show Plot
#```{r}
img <- readPNG("../figures/IMG_NumOrigins.png")
grid.raster(img)
```

# Cell Cycle Gene Pathways
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$C)
Cell.pathway <- temp[grep("Cell cycle", temp)]
Cell.Table <- gene.dictionary[which(gene.dictionary$C %in% Cell.pathway), ]
Cell.Table.S <- Cell.Table[which(Cell.Table$ko %in% single), ]
Cell.Genes <- Cell.Table.S$ko
Cell.Names <- Cell.Table.S$gene

Cell.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% Cell.Genes), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = Cell.Origins$ASR.Mean, y = 1:dim(Cell.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(Cell.Origins)[1] + 0.5))
arrows(x0 = Cell.Origins$ASR.Mean, 
       x1 = Cell.Origins$ASR.Mean + Cell.Origins$ASR.SE, 
      y0 = 1:dim(Cell.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = Cell.Origins$ASR.Mean, 
       x1 = Cell.Origins$ASR.Mean - Cell.Origins$ASR.SE, 
      y0 = 1:dim(Cell.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = Cell.Origins$ASR.Mean, y = 1:dim(Cell.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)


arrows(x0 = Cell.Origins$CON.Mean, 
       x1 = Cell.Origins$CON.Mean + Cell.Origins$CON.SE, 
      y0 = 1:dim(Cell.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = Cell.Origins$CON.Mean, 
       x1 = Cell.Origins$CON.Mean - Cell.Origins$CON.SE, 
      y0 = 1:dim(Cell.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = Cell.Origins$CON.Mean, y = 1:dim(Cell.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(Cell.Origins$Name), 
     at = 1:dim(Cell.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(Cell.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(Cell.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(Cell.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(Cell.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(Cell.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-Cell.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(Cell.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-Cell.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(Cell.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```

# Nitrogen Genes
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$C)
temp.pathway <- temp[grep("Nitrogen", temp)]
temp.Table <- gene.dictionary[which(gene.dictionary$C %in% temp.pathway), ]
temp.Table.S <- temp.Table[which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)

abline(v = 2450, lty = 2, lwd = 1, col = "gray")
abline(v = 1850, lty = 2, lwd = 1, col = "gray")
text(x = 2150, y = dim(temp.Origins)[1] + 0.65, "Great Oxidation")


arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```

# Photosynthesis Genes
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$gene)
temp.pathway <- temp[grep("ps[a,b]", temp)]
temp.Table <- gene.dictionary[which(gene.dictionary$gene %in% temp.pathway), ]
temp.Table.S <- temp.Table[which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)

abline(v = 2450, lty = 2, lwd = 1, col = "gray")
abline(v = 1850, lty = 2, lwd = 1, col = "gray")
text(x = 2150, y = dim(temp.Origins)[1] + 0.65, "Great Oxidation")

arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```


# Sulfur Genes
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$C)
temp.pathway <- c(temp[grep("Sulfur", temp)])
temp.Table <- gene.dictionary[which(gene.dictionary$C %in% temp.pathway), ]
temp.Table.S <- temp.Table[which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp <- unique(gene.dictionary$gene)
temp.pathway <- c(temp[grep("dms[A-Z]", temp)], temp[grep("hyd[A-Z]; sul", temp)],
                  temp[grep("dimethylsulfide", temp)], temp[grep("dox[A-Z]", temp)])
temp.Table <- gene.dictionary[which(gene.dictionary$gene %in% temp.pathway), ]
temp.Table.S <- temp.Table#[which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 8, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)

abline(v = 2450, lty = 2, lwd = 1, col = "gray")
abline(v = 1850, lty = 2, lwd = 1, col = "gray")
text(x = 2150, y = dim(temp.Origins)[1] + 0.5, "Great Oxidation")

arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 6, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```


# Methane Genes
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$C)
temp.pathway <- temp[grep("Methane", temp)]
temp.Table <- gene.dictionary[which(gene.dictionary$C %in% temp.pathway), ]
temp.Table.S <- temp.Table[which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)

abline(v = 2450, lty = 2, lwd = 1, col = "gray")
abline(v = 1850, lty = 2, lwd = 1, col = "gray")
text(x = 2150, y = dim(temp.Origins)[1] + 0.65, "Great Oxidation")


arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```

# Propanoate Genes
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$C)
temp.pathway <- temp[grep("Propanoate", temp)]
temp.Table <- gene.dictionary[which(gene.dictionary$C %in% temp.pathway), ]
temp.Table.S <- temp.Table[which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)

abline(v = 2450, lty = 2, lwd = 1, col = "gray")
abline(v = 1850, lty = 2, lwd = 1, col = "gray")
text(x = 2150, y = dim(temp.Origins)[1] + 0.65, "Great Oxidation")


arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```

# Benzoate Genes
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$C)
temp.pathway <- temp[grep("Benzoate degradation", temp)]
temp.Table <- gene.dictionary[which(gene.dictionary$C %in% temp.pathway), ]
temp.Table.S <- temp.Table[which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)

abline(v = 2450, lty = 2, lwd = 1, col = "gray")
abline(v = 1850, lty = 2, lwd = 1, col = "gray")
text(x = 2150, y = dim(temp.Origins)[1] + 0.65, "Great Oxidation")


arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```


# Urea Transport Genes
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$gene)
temp.pathway <- temp[grep("urt[A-Z]", temp)]
temp.Table <- gene.dictionary[which(gene.dictionary$gene %in% temp.pathway), ]
temp.Table.S <- temp.Table[which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)


arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```


# Urea Genes
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$gene)
temp.pathway <- temp[grep("ur[e,t][A-Z]", temp)]
temp.Table <- gene.dictionary[which(gene.dictionary$gene %in% temp.pathway), ]
temp.Table.S <- temp.Table# [which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)

abline(v = 2450, lty = 2, lwd = 1, col = "gray")
abline(v = 1850, lty = 2, lwd = 1, col = "gray")
text(x = 2150, y = dim(temp.Origins)[1] + 0.65, "Great Oxidation")


arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```

# Nitrogen Genes
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$gene)
temp.pathway <- temp[grep("n[a,i,o][r,k,p,s][A-Z]", temp)]
temp.Table <- gene.dictionary[which(gene.dictionary$gene %in% temp.pathway), ]
temp.Table.S <- temp.Table# [which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]
temp.Origins$Name


temp.Origins2 <- temp.Origins[grep("n[arG, apA, irK, irS, orB, osZ]", temp.Origins$Name), ]

temp.Origins$Name[grep("n[irS,irK]", temp.Origins$Name)]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)

abline(v = 2450, lty = 2, lwd = 1, col = "gray")
abline(v = 1850, lty = 2, lwd = 1, col = "gray")
text(x = 2150, y = dim(temp.Origins)[1] + 0.65, "Great Oxidation")



arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```

# Nitrogen Genes
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$gene)
temp.pathway <- temp[grep("ps[a,b][A-Z];", temp)]
temp.Table <- gene.dictionary[which(gene.dictionary$gene %in% temp.pathway), ]
temp.Table.S <- temp.Table[which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)

abline(v = 2450, lty = 2, lwd = 1, col = "gray")
abline(v = 1850, lty = 2, lwd = 1, col = "gray")
text(x = 2150, y = dim(temp.Origins)[1] + 0.85, "Great Oxidation")


arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```


# Oxygenase Genes
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$gene)
temp.pathway <- temp[grep("EC:1.14", temp)]
temp.Table <- gene.dictionary[which(gene.dictionary$gene %in% temp.pathway), ]
temp.Table.S <- temp.Table#[which(temp.Table$ko %in% single), ]
temp.Genes <- temp.Table.S$ko
temp.Names <- temp.Table.S$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)

abline(v = 2450, lty = 2, lwd = 1, col = "gray")
abline(v = 1850, lty = 2, lwd = 1, col = "gray")
text(x = 2150, y = dim(temp.Origins)[1] + 0.85, "Great Oxidation")


arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)
```


# Oxidoreductases
```{r}
multi <- names(which(table(gene.dictionary$ko) > 1))
single <- names(which(table(gene.dictionary$ko) == 1))
temp <- unique(gene.dictionary$gene)
temp.pathway <- temp[grep("EC:1.[1-21].", temp)]
temp.Table <- gene.dictionary[which(gene.dictionary$gene %in% temp.pathway), ]

#temp.Table.S <- temp.Table#[which(temp.Table$ko %in% single), ]

temp.Genes <- temp.Table$ko
temp.Names <- temp.Table$gene

temp.Origins <- Gene.Origins[which(Gene.Origins$Gene %in% temp.Genes), ]

temp.Origins <- temp.Origins[-c(which(is.na(temp.Origins$X))), ]

layout(matrix(1:2, 1, 2), widths = c(3, 1))
par(mar = c(5, 6, 2, 0))

plot(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 3, type = "n",
     axes = F, xlab = "", ylab = "",
     xlim = c(4000, 0), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean + temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
arrows(x0 = temp.Origins$ASR.Mean, 
       x1 = temp.Origins$ASR.Mean - temp.Origins$ASR.SE, 
      y0 = 1:dim(temp.Origins)[1], angle = 90, length = 0.05, lwd = 1.5)
points(x = temp.Origins$ASR.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5)

abline(v = 2450, lty = 2, lwd = 1, col = "gray")
abline(v = 1850, lty = 2, lwd = 1, col = "gray")
text(x = 2150, y = dim(temp.Origins)[1] + 0.85, "Great Oxidation")


arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean + temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
arrows(x0 = temp.Origins$CON.Mean, 
       x1 = temp.Origins$CON.Mean - temp.Origins$CON.SE, 
      y0 = 1:dim(temp.Origins)[1], 
      angle = 90, length = 0.05, lwd = 1.5, col = "red")
points(x = temp.Origins$CON.Mean, y = 1:dim(temp.Origins)[1], 
     pch = 22, bg = "gray", cex = 2, lwd = 1.5, col = "red")

axis(side = 1, labels = T, lwd.ticks = 1.5)
axis(side = 2, labels = c(temp.Origins$Name), 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 4, labels = F, 
     at = 1:dim(temp.Origins)[1], las = 1, lwd.ticks = 1.5)
axis(side = 2, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
axis(side = 4, labels = F, at = 1:dim(temp.Origins)[1], 
     tck = 0.02, lwd.ticks = 1.5)
mtext("Gene", side = 2, line = 4, cex = 1.5)
mtext("Age (mya)", side = 1, line = 3, cex = 1.5)
box(lwd = 1.5)

par(mar = c(5, 0, 2, 1))
plot(0, axes = F, xlab = "", ylab = "", type = "n",
     xlim = c(0, 1), ylim = c(0.5, dim(temp.Origins)[1] + 0.5))
text(x = 0.5, y = 1:dim(temp.Origins)[1]+0.2, cex = 0.75,
     labels = as.character(sprintf("%.4f", round(-temp.Origins$X * 1000, 4))))
text(x = 0.5, y = 1:dim(temp.Origins)[1]-0.2, cex = 0.75, 
     labels = as.character(sprintf("%.4f", round(-temp.Origins$Y * 1000, 4))))
text(x = 0.5, y = dim(temp.Origins)[1] + 0.75,
     labels = "Rates (X/Y); x 1000", xpd = T)


```






